src/nomad_simulations/schema_packages/basis_set.py

```diff
@@ -0,0 +1,642 @@
+        present_quantities = set(quantities) & self._get_open_quantities()
+        return [len(getattr(self, quant)) for quant in present_quantities]
+
+    def _of_equal_length(self, lengths: list[int]) -> bool:
+        """Check if all elements in the list are of equal length."""
+        if len(lengths) == 0:
+            return True
+        else:
+            ref_length = lengths[0]
+            return all(length == ref_length for length in lengths)
+
+    def get_n_terms(
+        self,
+        representative_quantities: set[str] = {
+            'energy_parameter',
+            'energy_parameter_n',
+            'differential_order',
+        },
+    ) -> Optional[int]:
+        """Determine the value of `n_terms` based on the lengths of the representative quantities."""
+        lengths = self._get_lengths(representative_quantities)
+        if not self._of_equal_length(lengths) or len(lengths) == 0:
+            return None
+        else:
+            return lengths[0]
+
+    def normalize(self, archive: EntryArchive, logger: BoundLogger) -> None:
+        super().normalize(archive, logger)
+
+        # enforce quantity length (will be used for type assignment)
+        new_n_terms = self.get_n_terms()
+        if self.n_terms is None:
+            self.n_terms = new_n_terms
+        elif self.n_terms != new_n_terms:
+            if logger is not None:
+                logger.error(
+                    f'Inconsistent lengths of `APWBaseOrbital` quantities: {self.m_def.quantities}. Setting back to `None`.'
+                )
+            self.n_terms = None
+
+        # enforce differential order constraints
+        if np.any(np.isneginf(self.differential_order)):
+            logger.error(
+                '`APWBaseOrbital.differential_order` must be completely non-negative.'
+            )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # it's hard to enforce commutative diagrams between `_determine_apw` and `normalize`
+        # instead, make all `_determine_apw` soft-coupled and dependent on the normalized state
+        # leverage normalize ∘ normalize = normalize
+        self._is_normalized = False
+
+
+class APWOrbital(APWBaseOrbital):
+    """
+    Implementation of `APWWavefunction` capturing the foundational (S)(L)APW basis sets, all of the form $\sum_{lm} \left[ \sum_o c_{lmo} \frac{\partial}{\partial r}u_l(r, \epsilon_l) \right] Y_lm$.
+    The energy parameter $\epsilon_l$ is always considered fixed during diagonalization, opposed to the original APW formulation.
+    This representation then has to match the plane-wave $k_n$ points within the muffin-tin sphere.
+
+    * D. J. Singh and L. Nordström, \"INTRODUCTION TO THE LAPW METHOD,\" in Planewaves, pseudopotentials, and the LAPW method, 2nd ed. New York, NY: Springer, 2006, pp. 43-52.
+    """
+
+    type = Quantity(
+        type=MEnum('apw', 'lapw', 'slapw'),  # ? where to put 'spherical_dirac'
```

Comment:
well, it is given as an option in the description. No idea what this is, so you can delete from the description or add it in the MEnum

Reply:
It's the relativistic correction to the basis set. I'd revisit this some other time.